Inconsistent Flow Rates in Swimming Pools with no Bypass Branches in Parallel

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #10102 
 - A bug was encountered by a user who had a swimming pool that did not have a bypass in parallel with it.  Essentially, the flow rate at the inlet and outlet of the swimming pool component were not equal which results in a lack of conservation of mass across the pool (makeup water is handled separately--the inlet and outlet water flow rate should always be the same).  When a bypass was present, this was not a problem because the flow resolver took the flow request of the pool and sent the remaining flow through the bypass.  However, when no bypass was present, it was determined that the loop was setting the mass flow rate based on the flow resolver and then this would be overwritten by the pool model, leading to a pool inlet and outlet flow rate that did not match.  The problem comes about because the swimming pool is unique in that it is present in the heat balance, in zone equipment, and directly in a plant loop.  When the flow resolver locks down the flow, this was not getting propagated correctly to the pool and then the temperatures were not being sent back to the plant loop.  To prevent this, when a swimming pool is present, the HVAC manager will take another pass through the various calls, even once the flow resolver locks everything down.  Testing with the defect .idf shows the flows at the inlet and outlet are correctly set by the loop in this situation and the outlet temperature is the appropriate value based on the pool energy balance.  When a bypass is present, the solution still manages that balance and does variable flow through the pool.  When no bypass or other parallel branch is present and flow is needed, all of the flow goes through the pool--resulting in an on-off operation based on the setpoint temperature and the pool water temperature.  A new unit test as well as minor documentation additions in both the Input-Output Reference and the Engineering Reference are included in this pull request.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Fix for Flow/Energy Inconsistencies

Swimming Pools under certain situations (no bypass) were producing results that were invalid.  Conservation of mass was not being preserved across the pool and neither was conservation of energy.  This was due to the swimming pool setting flows at the outlet but continuing to accept the inlet flow.  As a result of this, the inlet mass flow did not equal the exit mass flow.  In fact, in the defect file, it wasn't even close.  After much testing with different solutions, this solution fixed the situation where no bypass was present and did not seem to impact situations where a bypass was present (those results were fine already).

RKStrand committed:
Merge branch 'develop' into 10102-inconsistent-water-flow-rates-in-swimmingpoolindoor

RKStrand committed:
Pool Flow Issue Unit Test and Docs

Creation of a unit test to exercise the new subroutine that fixes the pool flow and energy lack of conservation in situations where there is no bypass. Mods to the IO Reference and Eng Ref to clarify flow rate issue.

rraustad commented on src/EnergyPlus/SwimmingPool.cc:
```diff
@@ -1181,4 +1181,22 @@ Real64 MakeUpWaterVolFunct(Real64 MakeUpWaterMass, Real64 Density)
+            thisPool.WaterMassFlowRate = maxMassFlowRate;
+        }
+    }
+}
```
Fundamentally, the equipment models should be asking the plant for a node mass flow rate. This call to SetComponentFlowRate should already be returning the proper water flow rate, at both the inlet and the outlet. 

    // Now calculate the requested mass flow rate from the plant loop to achieve the proper pool temperature
    // old equation using surface heat balance form: MassFlowRate = CpDeltaTi * ( CondTerms + ConvTerm + SWtotal + LWtotal + PeopleGain +
    // PoolMassTerm + MUWTerm + EvapEnergyLossPerArea );
    Real64 MassFlowRate = (this->WaterMass / (state.dataHVACGlobal->TimeStepSys * DataGlobalConstants::SecInHour)) *
                          ((TInSurf - TH22) / (TLoopInletTemp - TInSurf)); // Target mass flow rate to achieve the proper setpoint temperature
    if (MassFlowRate > this->WaterMassFlowRateMax) {
        MassFlowRate = this->WaterMassFlowRateMax;
    } else if (MassFlowRate < 0.0) {
        MassFlowRate = 0.0;
    }
    PlantUtilities::SetComponentFlowRate(state, MassFlowRate, this->WaterInletNode, this->WaterOutletNode, this->HWplantLoc);
    this->WaterMassFlowRate = MassFlowRate;

I think what you need here is a struct varible to keep track of what the pool mass flow rate request is this timestep. So before calling SetComponentFlowRate you would save this->WaterMassFlowRateRequest = MassFlowRate. Then call SetComponentFlowRate to update MassFlowRate, and then the difference between WaterMassFlowRateRequest and the node mass flow rate would be the bypass amount. It's seems this could be done all within the pool model instead of calling this new function multiple times from SimHVAC? And of course, I could be completely wrong on this so take this only as a suggestion. I also wonder why a pool even needs a bypass pipe since the pump sets the flow. I understand the model may want a certain flow, but it can only get what it gets.

RKStrand replied:
@rraustad Yes, I agree with you--it seems like this could all be done within the pool model.  I thought that what you suggested should have worked.  I tried that first or second I think (I tried several iterations on a similar theme) and for some reason it was not setting things correctly.  The SetComponentFlowRate call sets the flow request but that gets overwritten later in the flow resolver.  What was happening was, it would exit the pool routine with an outlet flow that matched the request and an outlet temperature that provided an energy balance at that flow.  Then, the flow resolver would come through and reset the outlet flow without touching the outlet temperature.  So, there was a net energy imbalance as a result.  Plus, the pool water flow rate would still throttle, leading to confusion for the user who logged the bug, because it didn't match the node flow rate.  I watched the flow rate and it would always get overwritten after the pool simulation step.  I also tried to query the flow lock variable within the swimming pool as a way to figure out when the flow was set, but that didn't work either.  It was at that point that I decided that I needed to take action at a higher level which led me to what I pushed.  I admit it's not ideal, but it's the only thing that I have tried that actually does what it is supposed to do.

On the topic of the pool getting a constant flow, that was also a point made by the individual who logged the bug.  The problem is that the variable flow algorithm is somewhat hard-wired into the routine.  Switching to a constant flow would involve some new equations for the main heat balance of the model.  While I understand the concern, that's more on the scope of a small development project.  Probably not a difficult one and also probably worth doing, but I'd consider it beyond a bug fix.  It would need at least one new input field to preserve the existing variable flow model.

You do make valid points and I would have guessed that what you proposed (like I said, one of the things that I tried early on) would have worked.  I guess my goal here was primarily to address the blatant energy imbalance and also clarify the flow rate, not try to create a new enhancement.

rraustad replied:
OK then, you tried doing things the correct way and that did not provide a solution. So be it. I do think that what you describe points to a bug in the plant flow resolver where the plant is updating nodes without components being aware of that change.

RKStrand replied:
That there could be a plant flow resolver bug is definitely a possibility.  My two thoughts were (1) wouldn't we see problems with other components too (it would seem like a swimming pool wouldn't be the only thing to see such a problem--hence my thought that it was something specific to the pool which uniquely bridges between the zone heat balances and a plant loop) and (2) the flow resolver seems pretty solid (I mean, it's been heavily vetted over many years, right?). I could be wrong, too.  I just know that I tried a variety of things that should have worked and none of those previous attempts did anything.

RKStrand replied:
@rraustad By the way, thanks for looking at this!

rraustad replied:
The other thing to be concerned with is setting an inlet or outlet node of a plant component that is inside a plant loop. Simply changing a node condition does not necessarily balance the entire plant. It might look good for the pool nodes, but are node conditions balanced througout the rest of the plant?

RKStrand replied:
Wow--well, that's a really good call.  What I thought was working from the perspective of the pool nodes is great.  I thought I checked the other nodes in my various test cases.  Apparently not.  The flows are fine, but the temperatures didn't get updated properly.  So, this isn't working yet...

RKStrand committed:
Revised solution

Revision to the solution previously proposed that wasn't actually working.  This works now and only adds to the simulation time when a pool is present.  The mass flow rate across the pool is now constant, even if a bypass is present.  So, the pool is now more like an on-off with full flow.  This was tested with the user defect .idf, a single branch demand loop variation of the user file, and a variation where a bypass was put in parallel with the pool.  All operate as constant flow/on-off control.

RKStrand committed:
Merge branch 'develop' into 10102-inconsistent-water-flow-rates-in-swimmingpoolindoor

RKStrand commented:
@rraustad I just made a commit that I think does actually resolve this issue now.  Hopefully it also addresses some of your concerns about the previous solution (which didn't work anyway).  Basically, the pool will now operate as a constant flow/on-off component not variable flow with bypass.  Now, having a bypass in parallel with the pool won't do anything.  Pools now satisfy continuity across the pool itself and downstream in the demand loop also.  After a lot of searching and head scratching, I realized that the plant was not updating after the pool simulation.  So, it would lock the flow but not propagate anything downstream of the pool, leaving both a mass and an energy imbalance.  That is resolved now by forcing the plant loops to resimulate when a pool is present.  This is actually necessary as the pool is the only component that is in zones, zone equipment, AND the plant.  With this extra simulation of the plant loops (again it's only done when pools are present) propagates the information now whereas before it wasn't.  This makes a lot more sense now (in my opinion).  Anyway, take a look at this and see what you think.  I still need to write a new unit test and circle back to the documentation which I'm pretty sure will need fixes in both the IO Ref and Eng Ref.  I'll make that commit once ci comes back relatively happy.  I expect differences in the existing Swimming Pool idf but hopefully there won't be any other diffs.

RKStrand committed:
Removed obsolete unit test

This unit test needs to be removed because it referenced a subroutine that got removed/replace.

RKStrand commented:
@rraustad Actually, I've backed out the changes within the swimming pool module itself and now I am very confident that what I am getting is actually correct.  This was a plant loop/flow resolver issue.  The reason for this is that the connection like the swimming pool was not considered when the flow lock rules were established.  The plant needs to be resimulated when there is a pool because even if the flow is locked, the outlet temperatures might get recalculated.  I anticipate the diffs for the two existing swimming pool models will be much smaller in the next commit that I am about to make and it will only go into on-off mode now when the pump forces it to.  This will satisfy the user who logged this and will still preserve the ability to run this as variable flow (if someone want to do that) using a defined bypass.  Still need to get a unit test together, but I wanted to keep you in the loop as to what is happening on this.

RKStrand committed:
Back out one more change

Backed out one more change that was made to the swimming pool model that actually is not needed.  The problem is really only in the way the flow resolver blocked the updating of information when a swimming pool was present.

RKStrand committed:
Solution to Pool Flow Issue

Modifications to the working solution, creation of a unit test, and final tweaks to the documentation prior to revirew.  This is ready to be looked at now.

RKStrand commented:
@rraustad Made a few more tweaks to the solution, added a new unit test, and clarified the solution in IO Ref and Eng Ref.  I will edit the PR description above shortly, but I think this is ready to be reviewed now.

Myoldmopar commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -853,7 +854,8 @@ void SimHVAC(EnergyPlusData &state)
     if (state.dataGlobal->AnyPlantInModel) {
-        if (PlantUtilities::AnyPlantSplitterMixerLacksContinuity(state)) {
+        if (PlantUtilities::AnyPlantSplitterMixerLacksContinuity(state) ||
+            state.dataSwimmingPools->anySwimmingPoolsInput) { // also need to resimulate when there are pools present
```
This feels....aggressive?  Maybe that's not the right word...specialized?  Niche?  

Why would a swimming pool be so different from other components like chillers or towers or ground loops?  I think the right course of action here is to simply ensure that the pool is resimulated as a plant component every single time plant runs.  If the outlet temperature changes, that's totally fine.  Let it find a solution.  Maybe I'm missing something?

Myoldmopar reviewed (commented):
I'm not quite convinced here.  Every time plant is simulated, the pool should determine how much flow it would like from plant, and use SetCompFlowRate.  Then it must use the flow plant gave it.  If the flowlock flag is unlocked, the plant will lodge the request and give it some flow.  If the flowlock flag is locked, it will simply give it whatever was resolved.  That's it.  I really just don't know about setting up a special flag for swimming pools when they should behave like every other component.

I could be off-base and missing something, let me know!

RKStrand committed:
Merge branch 'develop' into 10102-inconsistent-water-flow-rates-in-swimmingpoolindoor

Myoldmopar commented:
OK, I talked offline, and @RKStrand and I seem to agree that we should get this fixed in a more robust way, so I'm going to push this until after release.  I think the biggest issue is that we need the pool objects to exist as full plant components with a simulate() interface that is responsible for requesting a flow using SetComponentFlowRate.  If it does that, flow issues should go away, and any remaining problems would likely be due to loop restrictions or not setting the flow priority in the plant manager.

nrel-bot-2c commented:
@Myoldmopar @RKStrand @myoldmopar it has been 28 days since this pull request was last updated.
